### PR TITLE
[Model Monitoring] Implement TSDB abstraction

### PR DESCRIPTION
**Pull request by Eyal-Danieli**
Add Time Series DB abstraction that will be used for storing the model monitoring real time data. The PR includes:

- A new `tsdb` package under `mlrun/model_monitoring/db/tsdb`. The abstract class called `TSDBConnector` and it includes the main applications that are being used throughout the model monitoring processes.
- An implementation of `V3IOTSDBConnector` that is stored under `mlrun/model_monitoring/db/tsdb/v3io_connector`. Note that this object is based on V3IO `frames` client that handles the communications with V3IO TSDB. 
- The TSDB type is defined under a new configuration value called `tsdb_connector_type`. At the moment, the default value is `v3io-tsdb`.
- Create the TSDB application result table during the writer deployment (Instead of trying to generate it on every writer call).  

Related JIRAs - https://iguazio.atlassian.net/browse/ML-5795, https://iguazio.atlassian.net/browse/ML-5796

### CODE REVIEW

File: mlrun/model_monitoring/stream_processing.py [hunk 1]
```diff
@@ -321,77 +330,12 @@ def apply_storey_sample_window():
         # Steps 20-21 - Prometheus branch
         if not mlrun.mlconf.is_ce_mode():
```
Comment (davesh0812) [inline mlrun/model_monitoring/stream_processing.py:324]: I think you should move this check to `get_tsdb_connector` line 36 :)

File: mlrun/model_monitoring/db/tsdb/__init__.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    def to_tsdb_connector(self, project: str, **kwargs) -> TSDBConnector:
+        """
+        Return a TSDBConnector object based on the provided enum value.
+        :param project: The name of the project.
+        :return: `TSDBConnector` object.
+        """
+
+        if self == self.v3io_tsdb:
+            from .v3io.v3io import V3IOTSDBConnector
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/__init__.py]: I think its better to change the file name to be v3io.v3io_connector or v3io.connector

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        different key metric dictionaries. This data is being used by the monitoring dashboards in
+        grafana.
+        There are 3 different key metric dictionaries that are being generated throughout these steps:
+        - base_metrics (average latency and predictions over time)
+        - endpoint_features (Prediction and feature names and values)
+        - custom_metrics (user-defined metrics)
+        """
+        pass
+
+    def write_application_event(self, event: dict):
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py]:
```suggestion
    def write_application_result(self, event: dict):
```

better to be explicit because metrics will be write to  different tables.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        pass
+
+    def delete_tsdb_resources(self):
+        """
+        Delete all project resources in the TSDB connector, such as model endpoints data and drift results.
+        """
+        pass
+
+    def get_model_endpoint_real_time_metrics(
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py:75]: please move this function to crud as we talked.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+# limitations under the License.
+#
+
+
+from abc import ABC
+
+import pandas as pd
+
+
+class TSDBConnector(ABC):
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py:24]: Add 2 kinds of error one for raise from the `write` functions and one to raise from the `get_record` function, and maybe consider to add another one to the initialize function.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    def __init__(self, project: str):
+        """
+        Initialize a new TSDB connector.
+        :param project: the name of the project.
+        """
+        self.project = project
+
+    def apply_monitoring_stream_steps(self, graph):
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py:44]: Nice idea!!

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+from abc import ABC
+
+import pandas as pd
+
+
+class TSDBConnector(ABC):
+    def __init__(self, project: str):
+        """
+        Initialize a new TSDB connector.
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py]: Please add all the tables names (descriptions) that this class suppose to connect to.

File: mlrun/model_monitoring/writer.py [hunk 8]
```diff
@@ -153,30 +115,10 @@ def _update_kv_db(self, event: _AppResultEvent) -> None:
 
     def _update_tsdb(self, event: _AppResultEvent) -> None:
         event = _AppResultEvent(event.copy())
-        event[WriterEvent.END_INFER_TIME] = datetime.datetime.fromisoformat(
-            event[WriterEvent.END_INFER_TIME]
+        tsdb_connector = mlrun.model_monitoring.get_tsdb_connector(
```
Comment (davesh0812) [inline mlrun/model_monitoring/writer.py]: Please move to the `__init__()` method.
Also you can remove this function and to run `self.tsdb_connector.write_application_event(event=event)`
from the `do()` method.

Can be done also for the kv.

File: mlrun/model_monitoring/db/stores/v3io_kv/kv_store.py [hunk 9]
```diff
@@ -284,27 +283,23 @@ def delete_model_endpoints_resources(self, endpoints: list[dict[str, typing.Any]
...
-        store, _, _ = mlrun.store_manager.get_or_create_store(tsdb_path)
-        store.rm(tsdb_path, recursive=True)
+        tsdb_connector = mlrun.model_monitoring.get_tsdb_connector(
+            project=self.project,
+            access_key=self.access_key,
+            container=self.container,
+        )
+        tsdb_connector.delete_tsdb_resources()
+
+        if mlrun.mlconf.model_endpoint_monitoring.tsdb_connector_type == "v3io-tsdb":
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/stores/v3io_kv/kv_store.py]: I think it better to move this inside to `delete_tsdb_resources` of the v3io connector

File: mlrun/model_monitoring/db/stores/v3io_kv/kv_store.py [hunk 10]
```diff
@@ -284,27 +283,23 @@ def delete_model_endpoints_resources(self, endpoints: list[dict[str, typing.Any]
...
-        # Final cleanup of tsdb path
-        tsdb_path.replace("://u", ":///u")
-        store, _, _ = mlrun.store_manager.get_or_create_store(tsdb_path)
-        store.rm(tsdb_path, recursive=True)
+        tsdb_connector = mlrun.model_monitoring.get_tsdb_connector(
+            project=self.project,
+            access_key=self.access_key,
+            container=self.container,
+        )
+        tsdb_connector.delete_tsdb_resources()
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/stores/v3io_kv/kv_store.py]: Like discussed I think you need to had `endpoints` param to this function.
And also to call this function from `ModelEndpoints.delete_model_endpoints_resources()` and not from the kv/sql store.

File: mlrun/model_monitoring/db/tsdb/__init__.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    v3io_tsdb = "v3io-tsdb"
+
+    def to_tsdb_connector(self, project: str, **kwargs) -> TSDBConnector:
+        """
+        Return a TSDBConnector object based on the provided enum value.
+        :param project: The name of the project.
+        :return: `TSDBConnector` object.
+        """
+
+        if self == self.v3io_tsdb:
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/__init__.py:36]: check if not in ce mode :)

File: server/api/crud/model_monitoring/deployment.py [hunk 12]
```diff
@@ -242,6 +242,10 @@ def deploy_model_monitoring_writer_application(
                 writer_data=fn.to_dict(),
                 writer_ready=ready,
             )
+            # Create tsdb table for model monitoring application results
+            self._create_tsdb_application_tables(
+                project=fn.metadata.project, access_key=self.model_monitoring_access_key
```
Comment (davesh0812) [inline server/api/crud/model_monitoring/deployment.py]: No need to pass this, can cause a bug for user without admin permissions.
Because we are in the server side the env access key is all ready pipelines's.

File: server/api/crud/model_monitoring/deployment.py [hunk 13]
```diff
@@ -587,6 +591,27 @@ def is_monitoring_stream_has_the_new_stream_trigger(self) -> bool:
...
+        """Each project writer service writes the application results into a single TSDB table and therefore the
+        target table is created during the writer deployment"""
+
+        if (
+            mlrun.mlconf.model_endpoint_monitoring.tsdb_connector_type
+            == mm_constants.TSDBTarget.V3IO_TSDB
+        ):
+            tsdb_configurations = {
+                "access_key": kwargs["access_key"],
+            }
```
Comment (davesh0812) [inline server/api/crud/model_monitoring/deployment.py]: Because you don't need special access key this check is redundant, and you can use `get_tsdb_connector()` without extra kwargs.

**Review by davesh0812 — COMMENTED** [review]
Nice job! some comments, we discussed about most of themץ

File: mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+import mlrun.feature_store.steps
+from mlrun.common.schemas.model_monitoring import (
+    EventFieldType,
+    EventKeyMetrics,
+    EventLiveStats,
+)
+
+
+class ProcessBeforeTSDB(mlrun.feature_store.steps.MapClass):
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py:24]: We just duplicate the whole class from `stream_processing.py`?

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+import mlrun.utils.v3io_clients
+from mlrun.utils import logger
+
+_TSDB_BE = "tsdb"
+_TSDB_RATE = "1/s"
+
+
+class V3IOTSDBConnector(mlrun.model_monitoring.db.TSDBConnector):
+    """
+    Handles the TSDB operations when the TSDB connector is from type V3IO. To manage these operations we use V3IO Frames
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io.py]:
```suggestion
    Handles the TSDB operations when the TSDB connector is of type V3IO. To manage these operations we use V3IO Frames
```

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+        tsdb_batching_timeout_secs: int = 300,
+    ):
+        """
+        Apply TSDB steps on the provided monitoring graph. Throughout these steps, the graph stores live data of
+        different key metric dictionaries.This data is being used by the monitoring dashboards in
+        grafana. Results can be found under  v3io:///users/pipelines/project-name/model-endpoints/events/.
+        In that case, we generate 3 different key  metric dictionaries:
+        - base_metrics (average latency and predictions over time)
+        - endpoint_features (Prediction and feature names and values)
+        - custom_metrics (user-defined metrics
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io.py]:
```suggestion
        - custom_metrics (user-defined metrics)
```

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        """
+
+        # Step 12 - Before writing data to TSDB, create dictionary of 2-3 dictionaries that contains
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io.py]: It doesn't seem like this Step X stuff belong in the connector class, and could easily become wrong if someone adds a step earlier in the flow, elsewhere in the code.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+
+from abc import ABC
+
+import pandas as pd
+
+
+class TSDBConnector(ABC):
+    def __init__(self, project: str):
+        """
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py:26]: The constructor isn't normally documented, and the relevant information goes into the class docstring instead.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        columns: list[str] = None,
+        filter_query: str = "",
+        start: str = "now-1h",
+        end: str = "now",
+    ) -> pd.DataFrame:
+        """
+        Getting records from TSDB data collection.
+        :param table:            Path to the collection to query.
+        :param columns:          Columns to include in the result.
+        :param filter_query:     Filter expression.
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py]: That doesn't really explain what a filter expression/query is though.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        self,
+        table: str,
+        columns: list[str] = None,
+        filter_query: str = "",
+        start: str = "now-1h",
+        end: str = "now",
+    ) -> pd.DataFrame:
+        """
+        Getting records from TSDB data collection.
+        :param table:            Path to the collection to query.
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py]: Tables being located at paths is a rather V3IO-specific thing.

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py [hunk 21]
```diff
@@ -0,0 +1,365 @@
...
+    ) -> pd.DataFrame:
+        """
+         Getting records from V3IO TSDB data collection.
+        :param table:            Table key to query. The table key should exist in the class tables dictionary.
+        :param columns:          Columns to include in the result.
+        :param filter_query:     V3IO filter expression. The expected filter expression includes different conditions,
+                                 divided by ' AND '.
+        :param start:            The start time of the metrics. Can be represented by a string containing an RFC 3339
+                                 time, a Unix timestamp in milliseconds, a relative time (`'now'` or
+                                 `'now-[0-9]+[mhd]'`, where `m` = minutes, `h` = hours, and `'d'` = days), or 0 for the
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io.py]: `s` for second is supported as well.

File: mlrun/model_monitoring/stream_processing.py [hunk 22]
```diff
@@ -137,6 +137,12 @@ def _initialize_v3io_configurations(
         self.tsdb_batching_max_events = tsdb_batching_max_events
         self.tsdb_batching_timeout_secs = tsdb_batching_timeout_secs
 
+        self.tsdb_configurations = {
+            "access_key": self.v3io_access_key,
```
Comment (gtopper) [inline mlrun/model_monitoring/stream_processing.py]: Not specifically regarding the addition here, but this is no way to treat an access key...

**Review by gtopper — COMMENTED** [review]
Some comments. The main thing that troubles me about this PR though is that the abstraction is added with only one implementation (V3IO), so it remains to be seen if it would even hold for TDEngine.

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py [hunk 23]
```diff
@@ -0,0 +1,365 @@
...
+        for table in application_tables:
+            logger.info("Creating table in V3IO TSDB", table=table)
+            self._frames_client.create(
+                backend=_TSDB_BE,
+                table=self.tables[table],
+                if_exists=IGNORE,
+                rate=_TSDB_RATE,
+            )
+
+    def apply_monitoring_stream_steps(
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io.py]: Doesn't this function need to be broken up? It seems to contain some steps that are not related to V3IO.

Comment (gtopper) [reply mlrun/model_monitoring/db/tsdb/v3io/v3io.py]: I kind of expect the abstraction to be at the `apply_tsdb_target` level.

File: mlrun/model_monitoring/db/tsdb/base.py (hunk 7 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/base.py]: the thing is that we might have different tables structures across different TSDB solutions. Anyway, I'll add here a detailed description of the different types of monitoring data.

File: mlrun/model_monitoring/db/tsdb/base.py (hunk 4 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/base.py:75]: Since some of the functionality might be relatet to V3IO (such as how we process the df that is returned from get_records()), i kept it in the v3io connector. I did remove the relevant func from the kv_store and moved it into the crud part. 
Anyway, if we will see that this behaviour is repeating in other TSDBs (TDEngine) then I will move all of the logic to the crud part. Let me know if it's not clear

File: mlrun/model_monitoring/db/stores/v3io_kv/kv_store.py (hunk 10 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/stores/v3io_kv/kv_store.py]: as we agreed, in the meantime we won't pass endpoints to the delete_tsdb_resources. Anyway it delete all of the resources at the moment.

File: mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py (hunk 14 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py:24]: no, it's invalid :) I'll remove it from stream_processing.py

File: mlrun/model_monitoring/db/tsdb/base.py (hunk 19 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/base.py]: yeah you right but the thing is that the filter expression structure depends on the TSDB type (V3IO/TDEngine..). I'll try to document it better.

File: mlrun/model_monitoring/db/tsdb/base.py (hunk 20 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/base.py]: actually the docstring here is misleading. It's the table name (e.g. `app-results`, `events`..), not a path. I'll fix it.

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py (hunk 17 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/v3io/v3io.py]: Got it, I'll update the description and remove the "hard-coded" step numbering.

File: mlrun/model_monitoring/db/tsdb/v3io/v3io.py (hunk 23 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/v3io/v3io.py]: I'm not sure we are going to repeat these steps exactly in other TSDBs. Therefore, I preferred to keep it on the V3IO TSDB level at the moment.
 If for example we'll see that for TDEngine we need to do the exact logic then I'll move the "none-V3IO" steps into a higher level.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+        pass
+
+    def delete_tsdb_resources(self):
+        """
+        Delete all project resources in the TSDB connector, such as model endpoints data and drift results.
+        """
+
+        pass
+
+    def get_model_endpoint_real_time_metrics(
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py:75]: Please mention that those metrics are located and gets from the events table (at least for now).

File: server/api/crud/model_monitoring/deployment.py [hunk 25]
```diff
@@ -587,6 +589,23 @@ def is_monitoring_stream_has_the_new_stream_trigger(self) -> bool:
...
 
+    @staticmethod
+    def _create_tsdb_application_tables(project: str):
+        """Each project writer service writes the application results into a single TSDB table and therefore the
+        target table is created during the writer deployment"""
+
+        if (
+            mlrun.mlconf.model_endpoint_monitoring.tsdb_connector_type
+            == mm_constants.TSDBTarget.V3IO_TSDB
+        ):
```
Comment (davesh0812) [inline server/api/crud/model_monitoring/deployment.py]: Please remove this check, you already check it inside `get_tsdb_connector()`

File: server/api/crud/model_monitoring/model_endpoints.py [hunk 26]
```diff
@@ -546,18 +538,21 @@ def delete_model_endpoints_resources(project_name: str):
...
             ),
         )
-        endpoints = endpoint_store.list_model_endpoints()
+        endpoint_store.delete_model_endpoints_resources()
 
-        # Delete model endpoints resources from databases using the model endpoint store object
-        endpoint_store.delete_model_endpoints_resources(endpoints)
+        # Delete model monitoring TSDB resources
+        tsdb_connector = mlrun.model_monitoring.get_tsdb_connector(
+            project=project_name, access_key=auth_info.data_session
```
Comment (davesh0812) [inline server/api/crud/model_monitoring/model_endpoints.py]: again you don't need here to provide `access key` you are already on the server, and if the user that will run this api call won't have admin permissions this call will fail.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+
+import mlrun.common.schemas.model_monitoring.constants as mm_constants
+
+
+class TSDBConnector(ABC):
+    def __init__(self, project: str):
+        """
+        Initialize a new TSDB connector. The connector is used to interact with the TSDB and store monitoring data.
+        At the moment we have 3 different types of monitoring data:
+        - real time performance metrics: real time performance metrics that are being calculated by the model
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/base.py:29]:
```suggestion
        - real time performance metrics (named events): real time performance metrics that are being calculated by the model
```

File: mlrun/common/schemas/model_monitoring/constants.py [hunk 28]
```diff
@@ -228,6 +229,12 @@ class MonitoringFunctionNames(MonitoringStrEnum):
     WRITER = "model-monitoring-writer"
 
 
+class V3IOTSDBTables(MonitoringStrEnum):
```
Comment (davesh0812) [inline mlrun/common/schemas/model_monitoring/constants.py]: It spouse to be valid for all the tsdb's engines?  
```suggestion
class MonitoringTSDBTables(MonitoringStrEnum):
```

**Review by davesh0812 — APPROVED** [review]
LGTM.

Added few more comments :)

File: mlrun/db/httpdb.py [hunk 29]
```diff
@@ -3167,14 +3161,8 @@ def get_model_endpoint(
...
-        :param start:                      The start time of the metrics. Can be represented by a string containing an
-                                           RFC 3339 time, a Unix timestamp in milliseconds, a relative time (`'now'` or
-                                           `'now-[0-9]+[mhd]'`, where `m` = minutes, `h` = hours, and `'d'` = days), or
-                                           0 for the earliest time.
-        :param end:                        The end time of the metrics. Can be represented by a string containing an
-                                           RFC 3339 time, a Unix timestamp in milliseconds, a relative time (`'now'` or
-                                           `'now-[0-9]+[mhd]'`, where `m` = minutes, `h` = hours, and `'d'` = days), or
-                                           0 for the earliest time.
+        :param start:                      The start time of the metrics.
+        :param end:                        The end time of the metrics.
```
Comment (gtopper) [inline mlrun/db/httpdb.py]: Not really sure why all the information regarding these parameters was deleted?
(other occurrences as well)

File: mlrun/model_monitoring/db/tsdb/__init__.py [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        valid_values = list(cls.__members__.keys())
+        raise mlrun.errors.MLRunInvalidArgumentError(
+            f"{value} is not a valid tsdb, please choose a valid value: %{valid_values}."
+        )
+
+
+def get_tsdb_connector(project: str, **kwargs) -> TSDBConnector:
+    """
+    Getting the TSDB connector type based on mlrun.config.model_endpoint_monitoring.tsdb_connector_type.
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/__init__.py]:
```suggestion
    Get the TSDB connector type based on mlrun.config.model_endpoint_monitoring.tsdb_connector_type.
```

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+
+    def write_application_event(
+        self,
+        event: dict,
+        kind: mm_constants.WriterEventKind = mm_constants.WriterEventKind.RESULT,
+    ):
+        """
+        Write a single application or metric to TSDB.
+
+        :raise mlrun.errors.MLRunInvalidArgumentError: If an error occurred while writing the event.
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py:64]:
```suggestion
        :raises mlrun.errors.MLRunInvalidArgumentError: If an error occurred while writing the event.
```

Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py]: Are you sure this is the right exception to raise? Sounds more like an `MLRunRuntimeError`.

File: mlrun/model_monitoring/db/tsdb/base.py [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+        Getting records from TSDB data collection.
+        :param table:            Table name, e.g. 'metrics', 'app_results'.
+        :param columns:          Columns to include in the result.
+        :param filter_query:     Optional filter expression as a string. The filter structure depends on the TSDB
+                                 connector type.
+        :param start:            The start time of the metrics.
+        :param end:              The end time of the metrics.
+
+        :return: DataFrame with the provided attributes from the data collection.
+        :raise:  MLRunInvalidArgumentError if the provided table wasn't found.
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py:113]:
```suggestion
        :raises:  MLRunInvalidArgumentError if the provided table wasn't found.
```

Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/base.py]: This one sounds more like an `MLRunNotFoundError`.

File: mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py [hunk 33]
```diff
@@ -0,0 +1,117 @@
...
+            processed[EventKeyMetrics.CUSTOM_METRICS] = {
+                EventFieldType.RECORD_TYPE: EventKeyMetrics.CUSTOM_METRICS,
+                **event[EventFieldType.METRICS],
+                **base_event,
+            }
+
+        return processed
+
+
+class FilterAndUnpackKeys(mlrun.feature_store.steps.MapClass):
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py:92]: Is this meant to be duplicated verbatim from `mlrun/model_monitoring/stream_processing.py`?

File: mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py [hunk 34]
```diff
@@ -0,0 +1,407 @@
...
+        apply_tsdb_target(name="tsdb3", after="FilterNotNone")
+
+    def write_application_event(
+        self,
+        event: dict,
+        kind: mm_constants.WriterEventKind = mm_constants.WriterEventKind.RESULT,
+    ):
+        """
+        Write a single result or metric to TSDB.
+        """
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py]:
```suggestion
        """Write a single result or metric to TSDB"""
```

File: mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py [hunk 35]
```diff
@@ -0,0 +1,407 @@
...
+                self._frames_client.delete(
+                    backend=mlrun.common.schemas.model_monitoring.TimeSeriesConnector.TSDB,
+                    table=table,
+                )
+            except v3io_frames.errors.DeleteError as e:
+                if "No TSDB schema file found" not in str(e):
+                    logger.warning(
+                        f"Failed to delete TSDB table '{table}'",
+                        err=mlrun.errors.err_to_str(e),
+                    )
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py]: And if it's not, we just suppress the error entirely?

File: mlrun/db/httpdb.py (hunk 29 above)
Comment (Eyal-Danieli) [reply mlrun/db/httpdb.py]: I'll bring it back though my concern was that this information is only relevant to V3IO and this API is going to serve other TSDBs as well.

File: mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py (hunk 33 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/v3io/stream_graph_steps.py:92]: no ;/

File: mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py (hunk 35 above)
Comment (Eyal-Danieli) [reply mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py]: agree, it's an old `if` that seems quite redundant here. We need to log the err either way. I'll update it.

File: mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py [hunk 36]
```diff
@@ -0,0 +1,405 @@
...
+                                 `'s'` = seconds), or 0 for the earliest time.
+        :param end:              The end time of the metrics. Can be represented by a string containing an RFC 3339
+                                 time, a Unix timestamp in milliseconds, a relative time (`'now'` or
+                                 `'now-[0-9]+[mhd]'`, where `m` = minutes, `h` = hours, `'d'` = days, and
+                                 `'s'` = seconds), or 0 for the earliest time.
+        :return: DataFrame with the provided attributes from the data collection.
+        :raise:  MLRunNotFoundError if the provided table wasn't found.
+        """
+        if table not in self.tables:
+            raise mlrun.errors.MLRunInvalidArgumentError(
```
Comment (gtopper) [inline mlrun/model_monitoring/db/tsdb/v3io/v3io_connector.py]: Hm, which error is it? `MLRunNotFoundError` or `MLRunInvalidArgumentError`?

**Review by gtopper — APPROVED** [review]
